File: python/p17.py

Singles = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
Teens = ['eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen']
Tens = ['ten', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety']
Hundred = 'hundred'
Thousand = 'thousand'


def parseNum(num):
  thousands = ""
  hundreds = ""
  tens = ""
  singles = ""

  thousandsdigit = (num/1000)%10
  if thousandsdigit > 0:
    thousands = Singles[thousandsdigit-1] + Thousand

  hundredsdigit = (num/100)%10
  if hundredsdigit > 0:
    hundreds = Singles[hundredsdigit-1] + Hundred


  tensdigit = (num/10)%10
  onesdigit = num%10

  if hundredsdigit > 0 and (onesdigit > 0 or tensdigit > 0):
    hundreds += "and"

  if onesdigit > 0:
    singles = Singles[onesdigit-1]

  if tensdigit == 1 and onesdigit != 0:
    tens = Teens[onesdigit-1]
    singles = ""
  elif tensdigit >= 1:
    tens = Tens[tensdigit-1]
  
  return thousands + hundreds + tens + singles

lensum = 0
for x in range(1,1001):
  lensum += len(parseNum(x))
print lensum

File: c++/p17.cpp
#include "p17.hpp"
#include <string>

MAKE_PROTOTYPE(p17)

static const char *ones[] = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
static const char *tens[] = {"ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"};
static const char *teens[] = {"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"};
static const char *hundred = "hundred";

static std::string constructNumberWord(uint num)
{
  std::string result;
  uint numHund = num / 100;
  uint numTen = (num - numHund*100) / 10;
  uint numOnes = num - (numHund*100 + numTen*10);

  if ( numHund > 0 )
  {
    result += ones[numHund - 1];
    result += hundred;
    if ( numHund*100 < num )
      result += "and";
  }

  if ( numTen > 1 )
  {
    result += tens[numTen-1];
  }
  else if ( numTen == 1 )
  {
    result += teens[numOnes];
    return result;
  }
    
  if ( numOnes > 0 )
  {
    result += ones[numOnes - 1];
  }

  return result;
}

uint p17::execute() 
{
  uint result = 0;
  for ( uint i = 1; i < 1000; ++i )
  {
    result += constructNumberWord(i).size();
  }

  result += std::string("onethousand").size();

  return result;
}
